simtools/layout/telescope_position.py

```diff
@@ -17,597 +15,547 @@ class MissingInputForConvertion(Exception):
 
-    def getUtmCoordinates(self):
+    def _convert(self, crs_from, crs_to, xx, yy):
         """
-        Get utm north and east.
+        Coordinate transformation of telescope positions.
+        Returns np.nan for failed transformations (and not inf, as pyproj does)
+
+        Parameters
+        ----------
+        crs_from: pyproj.crs.crs.CRS
+            Projection of input data
+        crs_to: pyproj.crs.crs.CRS
+            Projection of output data
+        xx: scalar
+            Input x coordinate
+        yy: scalar
+            Input y coordinate
 
         Returns
         -------
-        (utmNorth [u.m], utmEast [u.m])
-        """
-        return self._utmNorth * u.m, self._utmEast * u.m
+        scalar
+            Output x coordinate
+        scalar
+            Output y coordinate
 
-    @u.quantity_input(utmEast=u.m, utmNorth=u.m)
-    def setUtmCoordinates(self, utmEast, utmNorth):
-        """Set the UTM coordinates."""
-        if None not in [self._utmEast, self._utmNorth]:
-            self._logger.warning("UTM coordinates are already set and will be overwritten")
+        Raises
+        ------
+        pyproj.exceptions.CRSError
+            If input or output projection (coordinate system) is not defined
 
-        self._utmEast = utmEast.value
-        self._utmNorth = utmNorth.value
+        """
+        try:
+            transformer = pyproj.Transformer.from_crs(crs_from, crs_to)
+        except pyproj.exceptions.CRSError:
+            self._logger.error("Invalid coordinate system")
+            raise
+        if xx is None or yy is None:
+            return np.nan, np.nan
+        _to_x, _to_y = transformer.transform(xx=xx, yy=yy)
+        if np.isinf(_to_x) or np.isinf(_to_y):
+            return np.nan, np.nan
+        return _to_x, _to_y
 
-    def hasUtmCoordinates(self):
+    def _get_reference_system_from(self):
         """
-        Return True if tel has UTM coordinates.
+        Return coordinate system and coordinatee for a fully defined system.
+        The firstfully defined system from self.crs is returned.
 
         Returns
         -------
-        bool
-        """
-        return self._utmEast is not None and self._utmNorth is not None
+        string
+            Name of coordinate system
+        pyproj.crs.crs.CRS
+            Project of coordinate system
 
-    def convertLocalToMercator(self, crsLocal, wgs84):
         """
-        Convert telescope position from local to mercator.
 
-        Parameters
-        ----------
-        crsLocal: pyproj.crs.crs.CRS
-            Pyproj CRS of the local coordinate system
-        wgs84: pyproj.crs.crs.CRS
-            Pyproj CRS of the mercator coordinate system
-
-        Raises
-        ------
-        InvalidCoordSystem
-            If crsLocal or wgs84 is not an instance of pyproj.crs.crs.CRS
-        """
+        for _crs_name, _crs in self.crs.items():
```

Comment:
I am aware that if several systems are given, it is hard to choose which one should be the reference one and simply the first one is chosen. What do you mean exactly by 'the reference system provided initially'?  One can initialize the class always with several systems, it doesn't have to be a single one.

Reply:
What I meant by the "reference system provided initially" is for example if I provide UTM positions to be converted to the other ones, the reference system should be UTM. That way we don't convert UTM --> CORSIKA --> UTM. I wrote this before I understood that technically the original coordinate system is provided to the layout array class, not to the telescope position class, but still, you do set the position of the telescope at least once. I think the first one provided should be used at the reference one, in order to avoid unnecessary conversions.

Reply:
Yes and no - I think it is a good idea to be absolute sure that all coordinates are consistent among itself after running convertAll. This is why I prefer to do the conversion of all systems.